rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java

```diff
@@ -441,33 +468,47 @@ private RMNode getNodebyUrl(String url) {
     }
 
     /**
-     * Set a node's state to free, after a completed task by it. Set the to free
-     * and move the node to the internal free nodes list. An event informing the
-     * node state's change is thrown to RMMonitoring.
+     * Change the state of the node to free, after a Task has been completed by the specified node.
+     * The node passed as parameter is moved to the list of nodes which are eligible for scheduling if it is not locked.
+     * In all cases, an event informing the node state's change is propagated to RMMonitoring.
      * 
      * @param rmNode node to set free.
      * @return true if the node successfully set as free, false if it was down before.
      */
     private BooleanWrapper internalSetFree(final RMNode rmNode) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("Current node state " + rmNode.getState() + " " + rmNode.getNodeURL());
+            logger.debug("Setting node state to free " + rmNode.getNodeURL());
+        }
+
         // If the node is already free no need to go further
-        logger.debug("Current node state " + rmNode.getState() + " " + rmNode.getNodeURL());
-        logger.debug("Setting node state to free " + rmNode.getNodeURL());
         if (rmNode.isFree()) {
             return new BooleanWrapper(true);
         }
+
         // Get the previous state of the node needed for the event
         final NodeState previousNodeState = rmNode.getState();
+
+
         Client client = rmNode.getOwner();
         if (client == null) {
             // node has been just configured, so the user initiated this action is the node provider
             client = rmNode.getProvider();
         }
-        // reseting owner here
+
+        // resetting owner here
         rmNode.setFree();
```

Comment:
When the node was previously locked, the node will be set in free state but not put in the freeNodes map. After looking at the code of other methods (lock, unlock). The way the node is handled is obscure to me. Would not a lockedNode collection be clearer to understand ?